### ISSUE TITLE
Improve speed / reduce memory allocations?

### ISSUE THREAD

**Issue by DavidBrust** [DavidBrust/FixedBed#1]
When running [simulations in 3D](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/PorousCatalystHot3D.jl) that also happen to include many function calls e.g. to determine physical properties, a considerable runtime is observed, especially for finer grids. A large number of allocations (in assembly loops) is reported by the solver.

I tried to pre-allocate some intermediate variables and store them in the "data" structure. I had limited success in speeding up the simulation (without chemical reactions) and also had convergence issuses (when including chemical reactions) [in this notebook.](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/FixAllocs.jl)
I might have made a mistake by pre-allocationg the intermediate variables as "Float64" or as Vector{Float64}.
What would be the correct way here?
What are possible strategies to speed up the simulation?

**Comment by j-fu** [comment]
Yes, I think the convergence problems come from not allocationg as dual numbers - while the code seems to run, it does not build the full Jacobi matrix which significantly slows down convergence even without reactions.

For fighting with allocations, we need two ingredients:
- mutable stack allocated arrays
- inplace matrix solution

After looking for the former I learnt that now StrideArrays.jl can provide this. For the later it is not complicated to write code (I did this e.g. for my lectures).

So I think this can be fixed. I e.g. may adapt one of your notebooks, and you can spread the method to the outhers.
I am currently at the SIAM CSE conference in Amsterdam, so I can promise this only for end of next week.

**Comment by j-fu** [comment]
BTW will you be at the MODVAL conference in Duisburg ?

**Comment by DavidBrust** [comment]
Thanks again for the information. I am looking forward to tackling this aspect.

I was not aware of the MODVAL conference before. It looks very interesting and relevant. Since registration is upon until Friday and it is nearby, I will try to attend :)

**Comment by j-fu** [comment]
I just looked at your code. May be you can sort things a little bit along the following lines before I would try to fix the allocation stuff:
- Returning data or tuples of data does not create allocations, so it is safe to write   `val1,val2=func(...)` no need to do `val1=zeros(1), val2=zeros(1); func!(val1,val2,...)
- Preallocating in the ModelData struct leads to intermediate storage  which ignores derivatives. So this should be avoided. The idea is to allocate intermediate storage in the flux functions using strided arrays.
- Kinetic parameters use Dicts with Strings as parameters. Better to replace the by Symbols: e.g. :CO2 instead of "CO2".  AFAIK this performs better and may even prevent allocations.

**Comment by j-fu** [comment]
Currently sitting in this session and getting more ideas: https://meetings.siam.org/sess/dsp_programsess.cfm?SESSIONCODE=75551 ...

**Comment by DavidBrust** [comment]
Thanks a lot for your suggestions, I will look into incorporating these.

Regarding the MODVAL conference: I will be attending :)

**Comment by DavidBrust** [comment]
I have incorporated the above mentioned changes into the [FixAllocations.jl](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/FixAllocations.jl) notebook.
Apart from the flux function, I also allocated intermediate storage with StrideArrays in the top function (calculating the top boundary condition) because the mixture composition and therefore molar fractions are required.

I hope this corresponds to what you had in mind, I unfortunately have not yet figured out, how to deal with allocations in Julia effectively...

**Comment by j-fu** [comment]
See #3 for an update.

**Comment by DavidBrust** [comment]
Thanks for your help and effort in avoiding allocations in the physics call backs!
I merged the proposed changes and was able to transfer the approach presented in #3 to kinetics.jl and the `reaction` callback, as can be seen in [FixAllocations.jl](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/FixAllocations.jl).

Unfortunately I struggle in doing the same to the 'main' model in [PorousCatalystHot3DTopFlowIrrExchange_NonAlloc.jl](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/PorousCatalystHot3DTopFlowIrrExchange_NonAlloc.jl), which is not converging.
The main difference to FixAllocations.jl is that it uses gas species enthalpies in the `flux` and `bcond` callbacks and more elaborate irradiation exchange with the environment.
The code in allocating form as it was before ( [PorousCatalystHot3DTopFlowIrrExchange.jl](https://github.com/DavidBrust/FixedBed/blob/main/notebooks/PorousCatalystHot3DTopFlowIrrExchange.jl) ) still converges, albeit slower than it used to prior to the changes to physprops.jl and kinetics.jl, that are conserved in the branch [Pre-alloc-fixing](https://github.com/DavidBrust/FixedBed/blob/Pre-alloc-fixing/notebooks/PorousCatalystHot3DTopFlowIrrExchange.jl)

**Comment by j-fu** [comment]
Just had a short look.
1) You should define
```
function enthalpy_mix(Fluids::AbstractVector, T, x)
    hmix = zero(eltype(x))
    ng=length(x)
    for i=1:ng
        hmix += x[i] * enthalpy_gas(Fluids[i], T)
    end
    hmix
end
```
i.e. add the AbstractVector here for proper dispatch.
After that the code was running on my computer.

2) For PorousCatalystHot3DTopFlowIrrExchange_NonAlloc.jl (and also for FixAllocations.jl, albeit not that seriously) we see stagnation of the contraction rate of Newton's method, i.e. we miss quadratic convergence. This is a sign that the Jacobi matrix is not calculated properly. The probable reason is that some intermediate data are stored as floats instead of duals.

**Comment by DavidBrust** [comment]
Thanks a lot for taking a look and your help.

Unfortunately, with my setup PorousCatalystHot3DTopFlowIrrExchange_NonAlloc.jl does not converge despite applying the `::AbstractVector` type declaration to the `enthalpy_mix` function. (the full problem including embedding does not converge, the initial solve converges, but only after 19 newton iterations) I also tried with Julia 1.9.2rc, which did not make any difference.

However, when I remove all calls to `enthalpy_mix` or `enthalpy_gas` from the `flux` and `bcond` callbacks, the newton method converges for the full problem.

**Comment by j-fu** [comment]
This seems to be connected with the conductiver temperature BC. I tried with reactions disabled. If one switches them off (in top,bottom), quadratic convergence is reached. I am  a bit puzzled why this is happening as I didn't see anything critical in that code. So it my be connected with the accuracy of Jacobian calculation. But then again this would be strange.
I guess I will try again... But in the moment it seems as if it would be worth to try to get everything else right. I plan to look into preconditioning and speed-up of assembly...
